File: src/engine/housekeeping.rs
use super::*;
use anyhow::Context;
use serde::Serialize;

#[derive(Serialize)]
struct StorageUsage {
    measured_at: String,
    application_bytes: u64,
    task_bytes: u64,
    planning_bytes: u64,
    runner_transcripts: Value,
}
impl App {
    pub(super) fn schedule_housekeeping(&self) {
        let now = chrono::Utc::now().timestamp();
        let (cleanup, observe) = {
            let mut rt = self.runtime();
            if rt.housekeeping.as_ref().is_some_and(|h| !h.is_finished()) {
                return;
            }
            rt.housekeeping = None;
            let cleanup = now - rt.last_retention_at >= 900;
            let observe = now - rt.last_observation_at >= 300;
            if !cleanup && !observe {
                return;
            }
            if cleanup {
                rt.last_retention_at = now;
            }
            if observe {
                rt.last_observation_at = now;
            }
            (cleanup, observe)
        };
        let app = self.clone();
        let handle = tokio::spawn(async move {
            let work = async {
                let c = app.config()?;
                if cleanup {
                    app.retention(&c).await?;
                    app.measure_storage(&c).await?;
                }
                if observe && !c.github_repo.is_empty() && c.repository.is_dir() {
                    app.observe_remote(&c).await?;
                }
                Ok::<_, anyhow::Error>(())
            }
            .await;
            if let Err(error) = work {
                let _ = app
                    .store
                    .event("system", "housekeeping_error", &format!("{error:#}"));
            }
        });
        self.runtime().housekeeping = Some(handle);
    }
    async fn measure_storage(&self, c: &Config) -> Result<()> {
        let dir = self.data_dir.clone();
        let roots = c.runner_storage_paths.clone();
        let usage=tokio::task::spawn_blocking(move || -> Result<_> {
            let mut runners=serde_json::Map::new();
            let mut total=0u64;
            let mut measured=0;
            for backend in ["codex","opencode"] {
                let (bytes,status)=match roots.get(backend) {
                    None=>(None,"unconfigured"),
                    Some(p) if !p.is_dir()=>(None,"unavailable"),
                    Some(p)=>match directory_size(p) {
                        Ok(b)=>(Some(b),"measured"),
                        Err(_)=>(None,"error"),
                    },
                };
                if let Some(b)=bytes { total=total.saturating_add(b); measured+=1; }
                runners.insert(backend.into(),json!({"bytes":bytes,"status":status}));
            }
            let status=if measured==runners.len(){"measured"}else if measured==0{"unavailable"}else{"partial"};
            Ok(StorageUsage { measured_at:now(),application_bytes:directory_size(&dir)?,
                task_bytes:directory_size(&dir.join("tasks"))?,planning_bytes:directory_size(&dir.join("cycles"))?,
                runner_transcripts:json!({"bytes":if measured>0{Some(total)}else{None},"status":status,"runners":runners,"message":"Runner storage reported separately. Application admission measures the data directory."}) })
        }).await??;
        self.store.put("settings", "storage", &usage)
    }
    async fn retention(&self, c: &Config) -> Result<()> {
        self.store.prune_events(c.retain_events)?;
        let cutoff = (chrono::Utc::now()
            - chrono::Duration::days(c.retain_completed_days.min(36500) as i64))
        .to_rfc3339();
        for mut check in self.store.baseline_cleanup_candidates()? {
            if self.shutdown.is_cancelled() {
                return Ok(());
            }
            {
                let _gate = self.gate.lock().await;
                let current: Option<crate::model::BaselineCheck> =
                    self.store.get("baseline", &check.id)?;
                let terminal = current
                    .as_ref()
                    .is_some_and(|c| c.status != crate::model::BaselineStatus::Running);
                let active = self
                    .runtime()
                    .baseline
                    .as_ref()
                    .is_some_and(|job| job.id == check.id);
                if !terminal || active {
                    continue;
                }
                if let Some(current) = current {
                    check = current;
                }
            }
            if let Err(error) = self.cleanup_baseline(&mut check).await {
                self.store
                    .event(&check.id, "cleanup_error", &format!("{error:#}"))?;
            }
        }
        for kind in ["task", "cycle"] {
            for id in self.store.cleanup_candidates(kind, &cutoff)? {
                if self.shutdown.is_cancelled() {
                    return Ok(());
                }
                let _gate = self.gate.lock().await;
                let result = if kind == "task" {
                    let mut task: Task =
                        self.store.get(kind, &id)?.context("Missing cleanup task")?;
                    if task.status.active() || self.runtime().tasks.contains_key(&id) {
                        continue;
                    }
                    self.discard_task(&mut task).await
                } else {
                    let mut cycle: Cycle = self
                        .store
                        .get(kind, &id)?
                        .context("Missing cleanup cycle")?;
                    if cycle.status == "running" {
                        continue;
                    }
                    self.discard_cycle(&mut cycle).await
                };
                if let Err(error) = result {
                    self.store
                        .event(&id, "cleanup_error", &format!("{error:#}"))?;
                }
            }
        }
        Ok(())
    }
    pub async fn discard_task(&self, task: &mut Task) -> Result<()> {
        ensure!(
            !task.status.active() && task.status != Status::Queued,
            "Active or queued workspaces cannot be discarded"
        );
        if !task.workspace.is_empty() {
            let path = Path::new(&task.workspace)
                .parent()
                .context("Invalid task workspace")?;
            let expected = self.task_workspace(&task.id);
            let expected = expected.parent().context("Invalid task workspace")?;
            let owner = path
                .file_name()
                .and_then(|s| s.to_str())
                .context("Invalid workspace owner")?;
            ensure!(
                path == expected || task.execution_session.as_deref() == Some(owner),
                "Cleanup path does not belong to this task or its legacy execution session"
            );
            remove_owned_dir(&self.data_dir.join("tasks"), path).await?;
        }
        task.lifecycle.discarded_at = Some(now());
        self.save_task(task)
    }
    pub async fn discard_cycle(&self, cycle: &mut Cycle) -> Result<()> {
        ensure!(
            cycle.status != "running",
            "Running planning work cannot be discarded"
        );
        uuid::Uuid::parse_str(&cycle.id).context("Invalid cycle workspace identity")?;
        remove_owned_dir(
            &self.data_dir.join("cycles"),
            &self.data_dir.join("cycles").join(&cycle.id),
        )
        .await?;
        cycle.lifecycle.discarded_at = Some(now());
        self.store.put("cycle", &cycle.id, cycle)
    }
    pub(crate) fn observe_pr(&self, c: &Config, p: PullRequest) -> Result<()> {
        self.persist_observation(c, p, false)
    }
    pub(crate) fn observe_delivery(&self, c: &Config, p: PullRequest) -> Result<()> {
        self.persist_observation(c, p, true)
    }
    fn persist_observation(&self, c: &Config, p: PullRequest, delivered_now: bool) -> Result<()> {
        let previous = self.store.pr_observation(&c.github_repo, p.number)?;
        let record_id = previous
            .as_ref()
            .map(|(id, _)| id.clone())
            .unwrap_or_else(|| format!("{}:{}", c.github_repo.to_ascii_lowercase(), p.number));
        let delivered = if delivered_now {
            Some(p.head.clone())
        } else if let Some(head) = previous.and_then(|(_, p)| p.delivered_head) {
            Some(head)
        } else {
            self.store.latest_pr_output(&c.github_repo, p.number)?
        };
        let observation = PrObservation {
            repository: c.github_repo.clone(),
            observed_at: now(),
            external_head_movement: delivered.as_ref().is_some_and(|head| head != &p.head),
            delivered_head: delivered,
            pr: p,
        };
        self.store.put("pr", &record_id, &observation)
    }
    async fn observe_remote(&self, c: &Config) -> Result<()> {
        let cancel = self.shutdown.child_token();
        git::validate_remote(c, &cancel).await?;
        let inventory = git::open_pr_inventory(c, &cancel).await?;
        for p in inventory.prs.iter().filter(|p| p.owned) {
            self.observe_pr(c, p.clone())?;
        }
        let open: HashSet<_> = inventory.prs.iter().map(|p| p.number).collect();
        // Poll known open PRs even when they disappear from open discovery results.
        let mut before = None;
        loop {
            let known = self.store.history_page(
                "pr",
                &crate::store::HistoryQuery {
                    before,
                    status: Some("open".into()),
                    limit: Some(100),
                    ..Default::default()
                },
            )?;
            for value in known.items {
                if self.shutdown.is_cancelled() {
                    return Ok(());
                }
                let number = value["pr"]["number"]
                    .as_u64()
                    .context("PR observation is missing its number")?;
                if value["repository"]
                    .as_str()
                    .is_some_and(|repository| repository.eq_ignore_ascii_case(&c.github_repo))
                    && !open.contains(&number)
                {
                    self.observe_pr(c, git::pr(c, number, &cancel).await?)?;
                }
            }
            before = known.next_cursor;
            if before.is_none() {
                break;
            }
        }
        self.reconcile_pr_inventory(c, &inventory, &cancel).await?;
        let observed_at = now();
        let revision = git::remote_revision(c, &c.default_branch, &cancel)
            .await?
            .unwrap_or_default();
        let fingerprint = context_fingerprint(&revision, &inventory.prs);
        if !revision.is_empty() {
            self.observe_default_branch(c, &revision, &observed_at)
                .await?;
        }
        let _gate = self.gate.lock().await;
        let live = self.config()?;
        if !live.github_repo.eq_ignore_ascii_case(&c.github_repo) {
            return Ok(());
        }
        let mut control = self.control()?;
        if !control.context_fingerprint.is_empty() && control.context_fingerprint != fingerprint {
            if control.idle_streak > 1 {
                // Reset an extended idle delay without bypassing the operator's
                // ordinary cadence (including changes from our own deliveries).
                control.next_cycle_at = control
                    .next_cycle_at
                    .min(chrono::Utc::now().timestamp() + c.cycle_interval_seconds as i64);
            }
            control.idle_streak = 0;
        }
        control.context_fingerprint = fingerprint;
        self.store.put("settings", "control", &control)
    }
}
pub(super) fn context_fingerprint(revision: &str, prs: &[PullRequest]) -> String {
    use sha2::{Digest, Sha256};
    let mut parts: Vec<_> = prs
        .iter()
        .map(|p| format!("{}:{}:{}:{}", p.number, p.head, p.base, p.state))
        .collect();
    parts.sort();
    format!(
        "{:x}",
        Sha256::digest(format!("{revision}\n{}", parts.join("\n")))
    )
}
pub(super) fn directory_size(path: &Path) -> Result<u64> {
    let entries = match std::fs::read_dir(path) {
        Ok(entries) => entries,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(0),
        Err(e) => return Err(e.into()),
    };
    let mut size = 0u64;
    for e in entries {
        let e = e?;
        let meta = match std::fs::symlink_metadata(e.path()) {
            Ok(m) => m,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => continue,
            Err(e) => return Err(e.into()),
        };
        if meta.is_symlink() {
            continue;
        }
        size = size.saturating_add(if meta.is_dir() {
            directory_size(&e.path())?
        } else {
            meta.len()
        });
    }
    Ok(size)
}
pub(super) async fn remove_owned_dir(root: &Path, path: &Path) -> Result<()> {
    ensure!(
        path.parent() == Some(root),
        "Cleanup path must be a direct child of the owned workspace root"
    );
    ensure!(
        path.file_name().is_some_and(|s| s != "." && s != ".."),
        "Invalid cleanup path"
    );
    for ancestor in path.ancestors() {
        match tokio::fs::symlink_metadata(ancestor).await {
            Ok(meta) => ensure!(!meta.is_symlink(), "Cleanup refuses symlink paths"),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
            Err(e) => return Err(e.into()),
        }
    }
    match tokio::fs::remove_dir_all(path).await {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(e.into()),
    }
}

File: internal/engine/housekeeping.go
package engine

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/tyk-swe/octomus-agent/internal/config"
	gitops "github.com/tyk-swe/octomus-agent/internal/git"
	"github.com/tyk-swe/octomus-agent/internal/model"
	"github.com/tyk-swe/octomus-agent/internal/store"
	"github.com/tyk-swe/octomus-agent/internal/workspace"
)

const (
	// retentionInterval and observeInterval pace the housekeeping passes.
	retentionInterval = 15 * time.Minute
	observeInterval   = 5 * time.Minute
	// observationLifetime is how long a remote observation (open-PR inventory
	// or default-branch revision) stays fresh for the dashboard. Each is
	// stamped only after its pass's earlier steps (retention, the storage walk,
	// the PR refresh), so a lifetime equal to observeInterval let a healthy
	// service flip to stale while the next, slower pass was still running.
	// Dispatch authority is separate and shorter (prAdmissionLifetime).
	observationLifetime = 2 * observeInterval
)

type storageUsage struct {
	MeasuredAt        string         `json:"measured_at"`
	ApplicationBytes  uint64         `json:"application_bytes"`
	TaskBytes         uint64         `json:"task_bytes"`
	PlanningBytes     uint64         `json:"planning_bytes"`
	RunnerTranscripts map[string]any `json:"runner_transcripts"`
}

// maybeStartHousekeeping may be called while gate is held. It only starts an
// owned background job; all filesystem and remote work happens after return.
func (a *App) maybeStartHousekeeping(cfg config.Config) {
	now := time.Now()
	a.runtimeMu.Lock()
	if a.runtime.housekeeping {
		a.runtimeMu.Unlock()
		return
	}
	cleanup := a.runtime.lastRetention.IsZero() || now.Sub(a.runtime.lastRetention) >= retentionInterval
	observe := a.runtime.lastObserve.IsZero() || now.Sub(a.runtime.lastObserve) >= observeInterval
	if !cleanup && !observe {
		a.runtimeMu.Unlock()
		return
	}
	a.runtime.housekeeping = true
	if cleanup {
		a.runtime.lastRetention = now
	}
	if observe {
		a.runtime.lastObserve = now
	}
	a.runtimeMu.Unlock()
	a.wg.Add(1)
	go func() {
		defer a.wg.Done()
		defer func() {
			a.runtimeMu.Lock()
			a.runtime.housekeeping = false
			a.runtimeMu.Unlock()
		}()
		// Once the service is stopping, the remaining steps are obsolete: the
		// pass ends at the next step boundary, and an interrupted step's
		// cancellation is not a housekeeping failure.
		report := func(err error) {
			if err != nil && a.ctx.Err() == nil {
				_ = a.Store.Event("system", "housekeeping_error", store.ErrorMessage(err))
			}
		}
		if cleanup {
			if err := a.retention(cfg); err != nil {
				report(err)
				return
			}
			if a.ctx.Err() != nil {
				return
			}
			if err := a.measureStorage(cfg); err != nil {
				report(err)
				return
			}
		}
		if a.ctx.Err() != nil {
			return
		}
		if stat, err := os.Stat(cfg.Repository); observe && cfg.GitHubRepo != "" && err == nil && stat.IsDir() {
			report(a.observeRemote(a.ctx, cfg))
		}
	}()
}

func (a *App) retention(cfg config.Config) error {
	if err := a.Store.PruneEvents(int64(cfg.RetainEvents)); err != nil {
		return err
	}
	days := cfg.RetainCompletedDays
	if days > 36500 {
		days = 36500
	}
	cutoff := time.Now().UTC().Add(-time.Duration(days) * 24 * time.Hour).Format(time.RFC3339)
	checks, err := a.Store.BaselineCleanupCandidates()
	if err != nil {
		return err
	}
	for _, check := range checks {
		if a.ctx.Err() != nil {
			return nil
		}
		// The gate covers only the eligibility re-read; CleanupBaseline claims
		// the check and removes its clone without the scheduler gate.
		a.gate.Lock()
		current, loadErr := store.Get[model.BaselineCheck](a.Store, "baseline", check.ID)
		terminal := current != nil && current.Status != model.BaselineStatusRunning
		a.runtimeMu.Lock()
		active := a.runtime.baseline != nil && a.runtime.baseline.id == check.ID
		a.runtimeMu.Unlock()
		a.gate.Unlock()
		if loadErr == nil && terminal && !active && current != nil {
			loadErr = a.CleanupBaseline(current)
		}
		if loadErr != nil {
			if eventErr := a.Store.Event(check.ID, "cleanup_error", store.ErrorMessage(loadErr)); eventErr != nil {
				return eventErr
			}
		}
	}
	for _, kind := range []string{"task", "cycle"} {
		ids, err := a.Store.CleanupCandidates(kind, cutoff)
		if err != nil {
			return err
		}
		for _, id := range ids {
			if a.ctx.Err() != nil {
				return nil
			}
			// The candidate list was read without the gate: the re-read under
			// it skips a record the operator discarded meanwhile, so its
			// discarded_at is never rewritten.
			a.gate.Lock()
			if kind == "task" {
				task, loadErr := store.Get[model.Task](a.Store, kind, id)
				if loadErr == nil && task != nil && task.Lifecycle.DiscardedAt == nil {
					a.runtimeMu.Lock()
					_, running := a.runtime.tasks[id]
					a.runtimeMu.Unlock()
					if !task.Status.Active() && !running {
						loadErr = a.DiscardTask(task)
					}
				}
				err = loadErr
			} else {
				cycle, loadErr := store.Get[model.Cycle](a.Store, kind, id)
				if loadErr == nil && cycle != nil && cycle.Lifecycle.DiscardedAt == nil && cycle.Status != model.CycleRunning {
					loadErr = a.DiscardCycle(cycle)
				}
				err = loadErr
			}
			a.gate.Unlock()
			// A conflict means another cleanup already owns this target —
			// success in progress, not a cleanup failure to report.
			if err != nil && !IsActionConflict(err) {
				if eventErr := a.Store.Event(id, "cleanup_error", store.ErrorMessage(err)); eventErr != nil {
					return eventErr
				}
			}
		}
	}
	return nil
}

// cleanupKind names the durable entity kind a cleanup claim owns. Each kind
// maps to one managed root, so the kind is part of the ownership unit.
type cleanupKind string

const (
	cleanupTask     cleanupKind = "task"
	cleanupCycle    cleanupKind = "cycle"
	cleanupBaseline cleanupKind = "baseline"
)

// cleanupKey identifies one managed-directory cleanup claim by durable entity
// kind and ID — each kind owns a distinct root, so the pair is the honest
// ownership unit (a path alone cannot distinguish a task from a cycle).
type cleanupKey struct {
	kind cleanupKind
	id   string
}

// claimCleanup takes exclusive cleanup ownership of (kind, id) and reports
// whether it was free. Callers claim while the scheduler gate is held so
// eligibility and ownership are one atomic admission — except CleanupBaseline,
// which claims first because its callers already serialized eligibility:
// retention re-reads the record under the gate and the finished check's worker
// owns its record. The map itself sits under runtimeMu.
func (a *App) claimCleanup(kind cleanupKind, id string) bool {
	a.runtimeMu.Lock()
	defer a.runtimeMu.Unlock()
	if a.runtime.cleanups == nil {
		a.runtime.cleanups = map[cleanupKey]struct{}{}
	}
	key := cleanupKey{kind: kind, id: id}
	if _, owned := a.runtime.cleanups[key]; owned {
		return false
	}
	a.runtime.cleanups[key] = struct{}{}
	return true
}

// cleanupClaimed reports whether (kind, id) is owned by an in-flight cleanup.
// Callers hold the scheduler gate; a claimed kind may be working gate-free
// under the baseline entry point (see claimCleanup).
func (a *App) cleanupClaimed(kind cleanupKind, id string) bool {
	a.runtimeMu.Lock()
	defer a.runtimeMu.Unlock()
	_, owned := a.runtime.cleanups[cleanupKey{kind: kind, id: id}]
	return owned
}

// releaseCleanup drops the claim. Releasing under the gate hold that wrote
// the final durable state leaves no gap between "removal finished" and
// "record marked" that a conflicting action could slip through.
func (a *App) releaseCleanup(kind cleanupKind, id string) {
	a.runtimeMu.Lock()
	delete(a.runtime.cleanups, cleanupKey{kind: kind, id: id})
	a.runtimeMu.Unlock()
}

// DiscardTask removes only the task's owned direct-child directory and marks
// the durable record after successful removal. Callers hold a.gate and get it
// back held: eligibility and the cleanup claim are checked under the gate,
// the recursive deletion runs with it released so unrelated controls stay
// responsive, and finalization re-reads the durable record so only the
// cleanup-owned field changes.
func (a *App) DiscardTask(task *model.Task) error {
	if task.Status.Active() || task.Status == model.StatusQueued {
		return errors.New("Active or queued workspaces cannot be discarded")
	}
	owner := ""
	if task.Workspace != "" {
		owner = filepath.Dir(filepath.Clean(task.Workspace))
		expected := filepath.Join(a.DataDir, "tasks", task.ID)
		if owner != expected {
			return errors.New("Cleanup path does not belong to this task")
		}
	}
	if !a.claimCleanup(cleanupTask, task.ID) {
		return conflictError("Workspace cleanup is already in progress for this task")
	}
	// Released on every exit, including removal failure and finalization
	// error, so a claim can never strand the record.
	defer a.releaseCleanup(cleanupTask, task.ID)
	if owner != "" {
		a.gate.Unlock()
		removeErr := a.removeDir(filepath.Join(a.DataDir, "tasks"), owner)
		a.gate.Lock()
		if removeErr != nil {
			return removeErr
		}
	}
	current, err := store.Get[model.Task](a.Store, "task", task.ID)
	if err != nil {
		return err
	}
	if current == nil {
		return nil
	}
	// A record that resumed work while the gate was released is not marked;
	// the claim normally prevents this, so treat it as an operator conflict.
	if current.Status.Active() || current.Status == model.StatusQueued {
		return conflictError("Task resumed work during workspace cleanup; inspect it before discarding")
	}
	now := model.Now()
	current.Lifecycle.DiscardedAt = &now
	if err := a.Store.Put("task", current.ID, *current); err != nil {
		return err
	}
	task.Lifecycle.DiscardedAt = current.Lifecycle.DiscardedAt
	return nil
}

// DiscardCycle removes a UUID-named planning directory and records disposal,
// under the same gate contract as DiscardTask: callers hold a.gate and the
// filesystem removal runs with it released.
func (a *App) DiscardCycle(cycle *model.Cycle) error {
	if cycle.Status == model.CycleRunning {
		return errors.New("Running planning work cannot be discarded")
	}
	if _, err := uuid.Parse(cycle.ID); err != nil {
		return errors.New("Invalid cycle workspace identity")
	}
	if !a.claimCleanup(cleanupCycle, cycle.ID) {
		return conflictError("Workspace cleanup is already in progress for this cycle")
	}
	defer a.releaseCleanup(cleanupCycle, cycle.ID)
	root := filepath.Join(a.DataDir, "cycles")
	a.gate.Unlock()
	removeErr := a.removeDir(root, filepath.Join(root, cycle.ID))
	a.gate.Lock()
	if removeErr != nil {
		return removeErr
	}
	current, err := store.Get[model.Cycle](a.Store, "cycle", cycle.ID)
	if err != nil {
		return err
	}
	if current == nil {
		return nil
	}
	if current.Status == model.CycleRunning {
		return conflictError("Planning work restarted during workspace cleanup; inspect it before discarding")
	}
	now := model.Now()
	current.Lifecycle.DiscardedAt = &now
	if err := a.Store.Put("cycle", current.ID, *current); err != nil {
		return err
	}
	cycle.Lifecycle.DiscardedAt = current.Lifecycle.DiscardedAt
	return nil
}

func (a *App) measureStorage(cfg config.Config) error {
	application, err := workspace.DirectorySize(a.DataDir)
	if err != nil {
		return err
	}
	tasks, err := workspace.DirectorySize(filepath.Join(a.DataDir, "tasks"))
	if err != nil {
		return err
	}
	planning, err := workspace.DirectorySize(filepath.Join(a.DataDir, "cycles"))
	if err != nil {
		return err
	}
	runners := map[string]any{}
	var total uint64
	measured := 0
	for _, backend := range []string{"codex", "opencode"} {
		path, configured := cfg.RunnerStoragePaths[backend]
		entry := map[string]any{"bytes": nil, "status": "unconfigured"}
		if configured {
			info, statErr := os.Stat(path)
			if statErr != nil || !info.IsDir() {
				entry["status"] = "unavailable"
			} else {
				bytes, sizeErr := workspace.DirectorySize(path)
				if sizeErr != nil {
					entry["status"] = "error"
				} else {
					entry["bytes"] = bytes
					entry["status"] = "measured"
					total += bytes
					measured++
				}
			}
		}
		runners[backend] = entry
	}
	status := "partial"
	var runnerBytes any = total
	if measured == 0 {
		status, runnerBytes = "unavailable", nil
	} else if measured == 2 {
		status = "measured"
	}
	usage := storageUsage{MeasuredAt: model.Now(), ApplicationBytes: application, TaskBytes: tasks, PlanningBytes: planning, RunnerTranscripts: map[string]any{"bytes": runnerBytes, "status": status, "runners": runners, "message": "Runner storage reported separately. Application admission measures the data directory."}}
	return a.Store.Put("settings", "storage", usage)
}

func (a *App) observeRemote(ctx context.Context, cfg config.Config) error {
	if err := gitops.ValidateRemote(ctx, cfg); err != nil {
		return err
	}
	if err := a.refreshPRs(ctx, cfg); err != nil {
		return err
	}
	inventory, err := a.Store.OpenPrInventory()
	if err != nil || inventory == nil {
		if err == nil {
			err = errors.New("PR refresh did not persist an inventory")
		}
		return err
	}
	open := map[uint64]struct{}{}
	for _, pr := range inventory.PRs {
		open[pr.Number] = struct{}{}
	}
	before := (*int64)(nil)
	status := "open"
	limit := 100
	closed := []model.PullRequest{}
	for {
		page, err := a.Store.HistoryPage("pr", store.HistoryQuery{Before: before, Status: &status, Limit: &limit})
		if err != nil {
			return err
		}
		for _, raw := range page.Items {
			if ctx.Err() != nil {
				return nil
			}
			var summary struct {
				Repository string `json:"repository"`
				PR         struct {
					Number uint64 `json:"number"`
				} `json:"pr"`
			}
			if json.Unmarshal(raw, &summary) != nil || !sameRepository(summary.Repository, cfg.GitHubRepo) {
				continue
			}
			if _, present := open[summary.PR.Number]; !present {
				pr, err := gitops.PR(ctx, cfg, summary.PR.Number)
				if err != nil {
					return err
				}
				closed = append(closed, pr)
			}
		}
		before = page.NextCursor
		if before == nil {
			break
		}
	}
	observedAt := model.Now()
	revision, err := gitops.RemoteRevision(ctx, cfg, cfg.DefaultBranch)
	if err != nil {
		return err
	}
	revisionValue := ""
	if revision != nil {
		revisionValue = *revision
	}
	fingerprint := ContextFingerprint(revisionValue, inventory.PRs)
	if revisionValue != "" {
		if err := a.observeDefaultBranch(cfg, revisionValue, observedAt); err != nil {
			return err
		}
	}
	a.gate.Lock()
	defer a.gate.Unlock()
	live, err := a.Config()
	if err != nil {
		return err
	}
	if !live.SameRemoteIdentity(cfg) {
		return nil
	}
	for _, pr := range closed {
		if err := a.Store.RecordPrObservation(cfg.GitHubRepo, pr, false); err != nil {
			return err
		}
	}
	control, err := a.Control()
	if err != nil {
		return err
	}
	if control.ContextFingerprint != "" && control.ContextFingerprint != fingerprint {
		if control.IdleStreak > 1 {
			ordinary := time.Now().Unix() + int64(cfg.CycleIntervalSeconds)
			if control.NextCycleAt > ordinary {
				control.NextCycleAt = ordinary
			}
		}
		control.IdleStreak = 0
	}
	control.ContextFingerprint = fingerprint
	return a.Store.SaveControl(control)
}

func ContextFingerprint(revision string, prs []model.PullRequest) string {
	parts := make([]string, 0, len(prs))
	for _, pr := range prs {
		parts = append(parts, fmt.Sprintf("%d:%s:%s:%s", pr.Number, pr.Head, pr.Base, pr.State))
	}
	sort.Strings(parts)
	return fmt.Sprintf("%x", sha256.Sum256([]byte(revision+"\n"+strings.Join(parts, "\n"))))
}
